sys/net/application_layer/dns/dns.c

```diff
@@ -73,55 +75,87 @@ static unsigned _get_short(uint8_t *buf)
     return _tmp;
 }
 
-static size_t _skip_hostname(uint8_t *buf)
+static ssize_t _skip_hostname(const uint8_t *buf, size_t len, uint8_t *bufpos)
 {
-    uint8_t *bufpos = buf;
+    const uint8_t *buflim = buf + len;
 
+    if (bufpos >= buflim) {
+        /* out-of-bound */
+        return -EBADMSG;
+    }
     /* handle DNS Message Compression */
     if (*bufpos >= 192) {
+        if ((bufpos + 2) >= buflim) {
+            return -EBADMSG;
+        }
         return 2;
     }
 
-    while(*bufpos) {
+    while (*bufpos) {
         bufpos += *bufpos + 1;
+        if (bufpos >= buflim) {
+            /* out-of-bound */
+            return -EBADMSG;
+        }
     }
     return (bufpos - buf + 1);
 }
 
 static int _parse_dns_reply(uint8_t *buf, size_t len, void* addr_out, int family)
 {
+    const uint8_t *buflim = buf + len;
     sock_dns_hdr_t *hdr = (sock_dns_hdr_t*) buf;
     uint8_t *bufpos = buf + sizeof(*hdr);
 
     /* skip all queries that are part of the reply */
     for (unsigned n = 0; n < ntohs(hdr->qdcount); n++) {
-        bufpos += _skip_hostname(bufpos);
-        bufpos += 4;    /* skip type and class of query */
+        ssize_t tmp = _skip_hostname(buf, len, bufpos);
+        if (tmp < 0) {
+            return tmp;
+        }
+        bufpos += tmp;
+        /* skip type and class of query */
+        bufpos += (RR_TYPE_LENGTH + RR_CLASS_LENGTH);
     }
 
     for (unsigned n = 0; n < ntohs(hdr->ancount); n++) {
-        bufpos += _skip_hostname(bufpos);
+        int tmp = _skip_hostname(buf, len, bufpos);
+        if (tmp < 0) {
+            return tmp;
+        }
+        bufpos += tmp;
+        if ((bufpos + RR_TYPE_LENGTH + RR_CLASS_LENGTH + RR_TTL_LENGTH) >= buflim) {
+            return -EBADMSG;
+        }
         uint16_t _type = ntohs(_get_short(bufpos));
-        bufpos += 2;
+        bufpos += RR_TYPE_LENGTH;
         uint16_t class = ntohs(_get_short(bufpos));
-        bufpos += 2;
-        bufpos += 4; /* skip ttl */
+        bufpos += RR_CLASS_LENGTH;
+        bufpos += RR_TTL_LENGTH; /* skip ttl */
 
         unsigned addrlen = ntohs(_get_short(bufpos));
-        bufpos += 2;
-        if ((bufpos + addrlen) > (buf + len)) {
-            return -EBADMSG;
-        }
-
         /* skip unwanted answers */
         if ((class != DNS_CLASS_IN) ||
                 ((_type == DNS_TYPE_A) && (family == AF_INET6)) ||
                 ((_type == DNS_TYPE_AAAA) && (family == AF_INET)) ||
                 ! ((_type == DNS_TYPE_A) || ((_type == DNS_TYPE_AAAA))
                     )) {
+            if ((bufpos + addrlen) < buf) {
+                /* buffer wraps around memory space */
+                return -EBADMSG;
+            }
```

Comment:
Thanks to _the amazing semantics of C_, this does not work as an overflow check, because pointer overflows are undefined behavior, allowing the compiler to both (a) omit this check because the condition will never be satisfied, and (b) allow `bufpos` to overflow with wrap-around in the next line. This was indeed done by GCC in the past: (CW: horrible techbros shouting at each other) https://gcc.gnu.org/bugzilla/show_bug.cgi?id=30475

Reply:
So how would *you* solve this?

Reply:
From what I know about optimization 6122e7258ada133e248fcdca5bf223a6b9623a6e *should* fix it.

Reply:
Just checking if ```addrlen > len``` should be sufficient to check if addrlen would cause overflow.

Reply:
As is, the second half of the expression creates undefined behaviour, too.

Reply:
I would rather not try to solve this. I tried doing overflow checks in the past, and it feels impossible. I stopped writing C instead, except for esotheric purposes.

What actually works in C is overflows of unsigned integers, which are guaranteed to wrap around.

I think it is valid to never change `buf`, remove `bufpos`, declare an index into the packet (must be unsigned) and use that to access the buffer contents (`buf[index]`), while advancing that index and comparing it to `len` (which also must be unsigned). That should only use defined behavior, I guess. One could define some fine accessor functions for reading unaligned 8, 16, and 32 bit values in network byte order from the buffer, while advancing the index and checking the bounds.

Reply:
> Just checking if `addrlen > len` should be sufficient to check if addrlen would cause overflow.

This works. addrlen can never be larger than len, and len would never cause overflow. The comparison os only using unsigned comparison. ([source](https://blog.regehr.org/archives/1395))

Reply:
Used @kaspar030's proposal.

Reply:
though... when incidentally `buf` is at the end of the RAM it might still wrap around with that :-/

Reply:
@pyropeter: If I remember correctly, regarding pointer overflows GCC decided to go back to sanity, as the used these type of checks in their own codebase as well. The link you posted is about overflows of signed integers.

Btw: Even more mind-blowing is if you change the code fefe posted to assign `a + 100` to a variable before doing the assertion.

```C
#include <assert.h>
#include <stdio.h>

int foo(int a) {
  int b = a + 100;
  assert(b > a);
  printf("%d %d\n",b,a);
  return a;
}

int main() {
  foo(100);
  foo(0x7fffffff);
}
```

Which still results in the output (for both GCC and clang):

```
200 100
-2147483549 2147483647
```

Without any assertion. Adding `-fwrapv` to `CFLAGS` works with both GCC and clang and turns sanity back on. Maybe we should consider adding `-fwrapv` to the `CFLAGS` then?

Reply:
> Adding -fwrapv to CFLAGS works with both GCC and clang and turns sanity back on. Maybe we should consider adding -fwrapv to the CFLAGS then?

@maribu do you want to open an issue or PR for that to discuss it?

Reply:
Sure. But today I have to do some teaching... I hope to get to it tomorrow.